drivers/timer/nrf_rtc_timer.c

```diff
@@ -63,7 +63,7 @@ void rtc1_nrf5_isr(void *arg)
 	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
 		u32_t next = last_count + CYC_PER_TICK;
 
-		if ((s32_t)(next - t) < MIN_DELAY) {
```

Comment:
I have a bad feeling about this...

The original code will add `CYC_PER_TICK` when t is past the next (negative number will be below the `MIN_DELAY`). So far I had a belief that this was the intention.
The new code will return high unsigned value and `CYC_PER_TICK` won't be added. If `t` is past `next` won't the next tick be triggered with high delay?

If I understand the change in this patch you try to fix overflow that may happen when `t` wrapped around `COUNTER_MAX`. Then indeed next will be high compared to `t` even if it should not be. But in this case shouldn't we check the overflow at the addition (`last_count + CYC_PER_TICK`)?

Reply:
Actually the original code would *fail* to add `CYC_PER_TICK` because `next - t` was a positive value greater than the 24-bit maximum counter value.  When it's truncated by `set_comparator` the effective counter (which produces a very small delay) might get stored too late and the wakeup be missed.

It doesn't matter if any value like `last_count + CYC_PER_TICK` (or `last_count += dticks * CYC_PER_TICK`) produces a value greater than 24-bits: the unsigned arithmetic operations "just work" for everything except for duration comparisons where we need to account for potential overflow/underflow.  (If the cast had been to `s24_t` it'd be fine, but that type doesn't exist.)

Reply:
Indeed @pdunaj spotted a silly place in the code. I agree that this should be fixed.

Reply:
to clarify ... I am not talking about the situation when we are close to the counter max but a situation when `t` is past `next`. i.e.
```
   last_count     next  t
-------------------------------------->time
```
however I noticed that `last_count` is updated from `t` in the irq so `next` cannot be behind `t`.

Reply:
@pizi-nordic I think you're concerned about the situation where the system clock ISR has been blocked for a long enough time that it gets serviced late, having "missed" ticks.

@pdunaj yes, exactly, `next` cannot be "behind" `t` because it's essentially `t` with a positive value added to it.  So the "missed" ticks are already accounted for in the adjustment to `last_count`.

It is possible there's something wrong.  But I'd want to see the specific situation clearly outlined with example values of `last_count`, `t`, `CYC_PER_TICK`, etc. to show what's going wrong.  I still think the code is correct as presented in the PR.

Reply:
@pabigot , @pizi-nordic , I overlooked `last_count` being updated few lines above. Code is fine and can be merged.